File: src/episode-update-upload.py
import csv
import os
import requests
import json
import mimetypes
import glob

# Constants
csv_file_path = './config/scotus.csv'
TRANSISTOR_API_URL = "https://api.transistor.fm/v1/episodes"
TRANSISTOR_API_KEY = os.getenv('TRANSISTOR_API_KEY')
TRANSISTOR_SHOW_ID = os.getenv('TRANSISTOR_SHOW_ID')
headers = {"x-api-key": TRANSISTOR_API_KEY, "Content-Type": "application/json"}
CASE_BASE_PATH = '../2023/'
IMGUR_CLIENT_ID = os.getenv('IMGUR_CLIENT_ID')

# Dry run flag
DRY_RUN = True


def find_first_file_in_directory(directory_path, file_types):
    """
    Finds the first file in the given directory that matches the file types.
    """
    for file_type in file_types:
        files = glob.glob(os.path.join(directory_path, f'*.{file_type}'))
        if files:
            return files[0]  # Return the first file found
    return None

def authorize_and_upload_audio(episode_id, audio_file_path):
    # Step 1: Authorize Audio Upload
    auth_url = f"{TRANSISTOR_API_URL}/{episode_id}/audio_upload"
    auth_response = requests.post(auth_url, headers=headers)
    upload_url = auth_response.json().get('data', {}).get('url')

    # Step 2: Upload Audio File
    with open(audio_file_path, 'rb') as audio_file:
        upload_response = requests.put(upload_url, data=audio_file)

    if upload_response.status_code != 200:
        print(f"Error uploading audio file: {upload_response.status_code}")
        return None

    return upload_url


def authorize_and_upload_image(episode_id, image_file_path):
    """
    Uploads an image to Imgur and returns the image URL.
    """
    headers = {
        'Authorization': f'Client-ID {IMGUR_CLIENT_ID}'
    }

    with open(image_file_path, 'rb') as image:
        data = {
            'image': image.read(),
            'type': 'file'
        }

        response = requests.post('https://api.imgur.com/3/upload', headers=headers, files=data)
        if response.status_code == 200:
            return response.json()['data']['link']
        else:
            print(f"Error uploading image: {response.status_code}")
            return None




def update_episode(episode_id, case_id, data):
    """
    Updates an episode in Transistor.fm with the given data, audio, and image.
    """
    audio_file = find_first_file_in_directory(os.path.join(CASE_BASE_PATH, case_id, 'audio'), ['mp3', 'wav'])
    image_file = find_first_file_in_directory(os.path.join(CASE_BASE_PATH, case_id, 'images'), ['jpg', 'jpeg', 'png'])

    audio_url = authorize_and_upload_audio(episode_id, audio_file) if audio_file else None
    image_url = authorize_and_upload_image(episode_id, image_file) if image_file else None  

    payload = {
        "episode": {
            "show_id": TRANSISTOR_SHOW_ID,
            "title": data['title'],
            "description": data['summary'],
            "audio_url": audio_url,  # Update the audio URL
            "image_url": image_url   # Update the image URL
        }
    }

    # If DRY_RUN is True, print the information instead of making an actual request
    if DRY_RUN:
        print(f"Dry run: Would update episode {episode_id} with data: {data}")
        print(f"Audio file to upload: {audio_file}")
        print(f"Image file to upload: {image_file}")
        # Close file handles if opened
        for file in files.values():
            file[1].close()
        return {"message": "Dry run, no update made"}

    # Make the actual request if not in dry run mode
    response = requests.put(url, headers=headers, data=payload, files=files)
    for file in files.values():
        file[1].close()

    if response.status_code != 200:
    print(f"Error updating episode: {response.status_code}")
    print(f"Response: {response.text}")  # More detailed logging

    return response.json()    



def process_csv(csv_path):
    """
    Processes each row in the CSV file and updates the corresponding episode.
    """
    with open(csv_path, mode='r', encoding='utf-8') as file:
        reader = csv.DictReader(file)

        for row in reader:
            if 'Episode ID' in row and row['Episode ID'] and 'Case ID' in row:
                episode_data = {
                    "title": row["Title"],
                    "summary": row["Summary"]
                }

                response = update_episode(row['Episode ID'], row['Case ID'], episode_data)
                print(f"Episode updated: {response}")

# Main execution
process_csv(csv_file_path)

File: src/go/episode-update-upload.go
package main

import (
    "bufio"
    "bytes"
    "encoding/csv"
    "encoding/json"
    "fmt"
    "io/ioutil"
    "mime/multipart"
    "net/http"
    "os"
    "path/filepath"
)

const (
    transistorAPIURL = "https://api.transistor.fm/v1/episodes"
    imgurAPIURL      = "https://api.imgur.com/3/upload"
    caseBasePath     = "../2023/"
    dryRun           = true
)

type EpisodeData struct {
    Title       string `json:"title"`
    Summary     string `json:"summary"`
    AudioURL    string `json:"audio_url,omitempty"`
    ImageURL    string `json:"image_url,omitempty"`
}

func findFirstFileInDirectory(directoryPath string, fileTypes []string) (string, error) {
    for _, fileType := range fileTypes {
        files, err := filepath.Glob(filepath.Join(directoryPath, fmt.Sprintf("*.%s", fileType)))
        if err != nil {
            return "", err
        }
        if len(files) > 0 {
            return files[0], nil
        }
    }
    return "", fmt.Errorf("no files found")
}

func authorizeAndUploadAudio(episodeID, audioFilePath string) (string, error) {
    // Step 1: Authorize Audio Upload
    authURL := fmt.Sprintf("%s/%s/audio_upload", transistorAPIURL, episodeID)
    req, err := http.NewRequest("POST", authURL, nil)
    if err != nil {
        return "", err
    }
    req.Header.Set("x-api-key", os.Getenv("TRANSISTOR_API_KEY"))
    req.Header.Set("Content-Type", "application/json")

    client := &http.Client{}
    authResp, err := client.Do(req)
    if err != nil {
        return "", err
    }
    defer authResp.Body.Close()

    var authResponse map[string]interface{}
    err = json.NewDecoder(authResp.Body).Decode(&authResponse)
    uploadURL, ok := authResponse["data"].(map[string]interface{})["url"].(string)
    if !ok {
        return "", fmt.Errorf("authorization failed")
    }

    // Step 2: Upload Audio File
    audioFile, err := os.Open(audioFilePath)
    if err != nil {
        return "", err
    }
    defer audioFile.Close()

    uploadResp, err := http.Post(uploadURL, "audio/mpeg", audioFile)
    if err != nil {
        return "", err
    }
    defer uploadResp.Body.Close()

    if uploadResp.StatusCode != http.StatusOK {
        return "", fmt.Errorf("error uploading audio file: %s", uploadResp.Status)
    }

    return uploadURL, nil
}

func authorizeAndUploadImage(episodeID, imageFilePath string) (string, error) {
    // Step 1: Prepare image file
    imageFile, err := os.Open(imageFilePath)
    if err != nil {
        return "", err
    }
    defer imageFile.Close()

    body := &bytes.Buffer{}
    writer := multipart.NewWriter(body)
    part, err := writer.CreateFormFile("image", filepath.Base(imageFilePath))
    if err != nil {
        return "", err
    }
    _, err = io.Copy(part, imageFile)
    if err != nil {
        return "", err
    }
    writer.Close()

    // Step 2: Upload Image
    req, err := http.NewRequest("POST", imgurAPIURL, body)
    if err != nil {
        return "", err
    }
    req.Header.Set("Authorization", fmt.Sprintf("Client-ID %s", os.Getenv("IMGUR_CLIENT_ID")))
    req.Header.Set("Content-Type", writer.FormDataContentType())

    client := &http.Client{}
    uploadResp, err := client.Do(req)
    if err != nil {
        return "", err
    }
    defer uploadResp.Body.Close()

    var uploadResponse map[string]interface{}
    err = json.NewDecoder(uploadResp.Body).Decode(&uploadResponse)
    imageURL, ok := uploadResponse["data"].(map[string]interface{})["link"].(string)
    if !ok {
        return "", fmt.Errorf("image upload failed")
    }

    return imageURL, nil
}

func updateEpisode(episodeID, caseID string, data EpisodeData) (json.RawMessage, error) {
    if dryRun {
        fmt.Printf("Dry run: Would update episode %s with data: %v\n", episodeID, data)
        return json.RawMessage(`{"message": "Dry run, no update made"}`), nil
    }

    payload, err := json.Marshal(map[string]EpisodeData{"episode": data})
    if err != nil {
        return nil, err
    }

    updateURL := fmt.Sprintf("%s/%s", transistorAPIURL, episodeID)
    req, err := http.NewRequest("PATCH", updateURL, bytes.NewBuffer(payload))
    if err != nil {
        return nil, err
    }
    req.Header.Set("x-api-key", os.Getenv("TRANSISTOR_API_KEY"))
    req.Header.Set("Content-Type", "application/json")

    client := &http.Client{}
    updateResp, err := client.Do(req)
    if err != nil {
        return nil, err
    }
    defer updateResp.Body.Close()

    var response json.RawMessage
    err = json.NewDecoder(updateResp.Body).Decode(&response)
    return response, err
}

func processCSV(csvPath string) error {
    file, err := os.Open(csvPath)
    if err != nil {
        return err
    }
    defer file.Close()

    reader := csv.NewReader(bufio.NewReader(file))
    rows, err := reader.ReadAll()
    if err != nil {
        return err
    }

    for _, row := range rows[1:] {
        episodeID := row[0] // Assuming 'Episode ID' is the first column
        caseID := row[1]   // Assuming 'Case ID' is the second column

        audioFile, err := findFirstFileInDirectory(filepath.Join(caseBasePath, caseID, "audio"), []string{"mp3"})
        if err != nil {
            fmt.Println("Error finding audio file:", err)
            continue
        }
        imageFile, err := findFirstFileInDirectory(filepath.Join(caseBasePath, caseID, "images"), []string{"jpg", "jpeg", "png"})
        if err != nil {
            fmt.Println("Error finding image file:", err)
            continue
        }

        audioURL, err := authorizeAndUploadAudio(episodeID, audioFile)
        if err != nil {
            fmt.Println("Error uploading audio:", err)
            continue
        }
        imageURL, err := authorizeAndUploadImage(episodeID, imageFile)
        if err != nil {
            fmt.Println("Error uploading image:", err)
            continue
        }

        episodeData := EpisodeData{
            Title:    row[2], // Assuming 'Title' is the third column
            Summary:  row[3], // Assuming 'Summary' is the fourth column
            AudioURL: audioURL,
            ImageURL: imageURL,
        }

        response, err := updateEpisode(episodeID, caseID, episodeData)
        if err != nil {
            fmt.Println("Error updating episode:", err)
            continue
        }
        fmt.Printf("Episode updated: %s\n", string(response))
    }
    return nil
}

func main() {
    csvFilePath := "./config/scotus.csv"

    if err := processCSV(csvFilePath); err != nil {
        fmt.Println("Error processing CSV file:", err)
    }
}
